keel-sql/src/main/kotlin/com/netflix/spinnaker/migrations/EventUidAssigner.kt

```diff
@@ -0,0 +1,73 @@
+package com.netflix.spinnaker.migrations
+
+import com.netflix.spinnaker.keel.persistence.metamodel.Tables.EVENT
+import com.netflix.spinnaker.keel.sql.inTransaction
+import de.huxhorn.sulky.ulid.ULID
+import java.time.LocalDateTime
+import java.time.ZoneOffset.UTC
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import org.jooq.DSLContext
+import org.jooq.Record1
+import org.slf4j.LoggerFactory
+import org.springframework.boot.context.event.ApplicationReadyEvent
+import org.springframework.context.event.EventListener
+
+class EventUidAssigner(private val jooq: DSLContext) : CoroutineScope {
+
+  private val idGenerator = ULID()
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  @EventListener(ApplicationReadyEvent::class)
+  fun onApplicationReady() {
+    launch {
+      while (countEventsWithNoUID() > 0) {
+        var count = 0
+        jooq.inTransaction {
+          select(EVENT.TIMESTAMP)
+            .from(EVENT)
+            .where(EVENT.UID.isNull)
+            .limit(25)
+            .forShare()
+            .fetch()
+            .forEach { ts ->
+              runCatching {
+                update(EVENT)
+                  .set(EVENT.UID, ts.nextULID())
+                  // this might actually not update the same row, but 🤷‍
+                  .where(EVENT.TIMESTAMP.eq(ts.value1()))
+                  .and(EVENT.UID.isNull)
+                  .limit(1)
+                  .execute()
+              }
+                .onSuccess { count += it }
+                .onFailure { ex ->
+                  log.error("Error assigning uid to event with timestamp ${ts.value1()}", ex)
+                }
```

Comment:
This updates a single row with a matching timestamp to the one we just selected with a new `uid`. It doesn't really matter if it's the same exact row. There's no good uniqueness as it stands, so it's practically impossible to ensure we are updating the same row. However, _most_ event timestamps are unique. The `limit(1)` ensures we always assign a unique uid.

Reply:
Yeah, that was pretty much the query I used when I had to do this before while testing locally. The only other idea I might offer would be to include `scope`, `json->'$.id'` and `json->'$.type'` in the criteria, which _should_ give you a single row.

Reply:
Yeah, ultimately it's not important though. Either way we'll end up with all the rows having a unique ULID.